snuba/clusters/cluster.py

```diff
@@ -134,6 +134,31 @@ def get_batch_writer(
 ClickhouseWriterOptions = Optional[Mapping[str, Any]]
 
 
+class ConnectionCache:
+    def __init__(self) -> None:
+        self.__cache: MutableMapping[
+            Tuple[ClickhouseNode, ClickhouseClientSettings], ClickhousePool
+        ] = {}
+
+    def __contains__(self, key: object) -> bool:
+        return key in self.__cache
+
+    def __getitem__(
+        self, key: Tuple[ClickhouseNode, ClickhouseClientSettings]
+    ) -> ClickhousePool:
+        return self.__cache[key]
+
+    def __setitem__(
+        self,
+        key: Tuple[ClickhouseNode, ClickhouseClientSettings],
+        value: ClickhousePool,
+    ) -> None:
+        self.__cache[key] = value
+
+
+connection_cache = ConnectionCache()
```

Comment:
My only concern with this global variable is that we use threading throughout the application. This object is not thread safe, are the ClickousePools threadsafe?

Reply:
I guess it is possible for us to set the cache key/value simultaneously in separate threads instead of reusing the same connection pool. Though this is also the case today and it seems relatively harmless. Are there more serious concerns I am not considering?

Reply:
ClickhousePool is thread safe. It uses `queue.LifoQueue` to store connections and it prevents them from being shared. This class indeed is not thread safe. I would synchronize the access just in case to avoid opening too many connections in case of race conditions.

Reply:
Ok updated to avoid this scenario by making the connection cache manage the connections. In doing so I realised there was another issue where it would assume the user/password/database was the same across clusters and reuse those connections even if that wasn't the case. That should now be fixed as part of this.

Reply:
This is still not thread safe as you can create more connections than you need, but better than before.

Reply:
seems pretty easy to put a `threading.Lock` around either this or the `__cache` member. Why not just do that?

Also, can we add a test to make sure that when the username/password are different, a different connection is made. I know it's obvious from reading the code now but we should enshrine that behavior with a test